fix(flags): Implement concurrency controls for feature flags

haacked opened the pull request:
## Problem

The current strategy for handling concurrency when editing feature flags is last one wins. This can lead to apparent data loss if two users are editing the same feature flag around the same time. The last person to save their changes wins, possibly overwriting changes from the first person.

There have been a couple of reports of this from customers:

_Internal tickets:_

* [posthoghelp.zendesk.com/agent/tickets/6500](https://posthoghelp.zendesk.com/agent/tickets/6500)
* [posthoghelp.zendesk.com/agent/tickets/24586](https://posthoghelp.zendesk.com/agent/tickets/24586)

Closes: #17995

## Changes

This PR implements a simple concurrency control in a similar fashion to how Notebooks do it. It adds a `version` field to `FeatureFlag`. This field is sent to the client, and when it's posted back, we compare it to the latest version in the database. If the versions match, we allow the update. If not, we reject it and suggest the user reload.

This also adds a `last_modified_by` column as a convenience so we can tell the user who was the last to modify it if they run into a conflict.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

* Manually tested it (see demo video https://www.loom.com/share/11a5416e7f1c412aa49362bc17da06a0?sid=5597e9fa-7259-4997-8aed-3426b5b7bb7d)
* Added unit tests

### Test Plan

- [x] Create a feature flag. Modify it. Ensure no conflict.
- [x] Create a feature flag. Modify it as a different user. Ensure no conflict.
- [x] Create a feature flag. Set `version` to null to simulate an existing flag. Modify it. Ensure no conflict.
- [x] In two browsers, edit a feature flag. Make changes. Save in one. Save in the other. Ensure conflict.
- [x] In two browsers, edit a feature flag. Make changes. Save in one. Refresh the other. Save. Ensure no conflict.
- [x] In two browsers, view the flag details page. User 1 clicks edit, saves changes. AFTERWARDS, user 2 clicks edit. Edit form should have updated values.

## Open questions and notes

* This PR does a dumb check. It's a bit more involved, but we could try and compare the changes to existing and only report a conflict if there are conflicting changes.
* We may want to track how often this happens to see if it's worth investing any more time. Is the right way to do that with logging or capturing an event?

haacked committed:
Add concurrency properties to `FeatureFlag`

This adds two new properties of `FeatureFlag`

* `version` - used to determine conflicts while editing a flag
* `last_modified_by` - the last user to modify a flag.

haacked committed:
Enforce concurrency

When modifying a feature flag, if the version has changed, we raise a conflict.

Fixes #17995

haacked committed:
Add assertion about the creator

haacked committed:
Add migration for feature flag concurrency columns

github-actions[bot] committed:
Update query snapshots

haacked committed:
Fix a test

haacked committed:
Allow null for the new field

haacked committed:
Handle null versions

Existing flags will have a null version, so lets make sure folks can update those too.

haacked committed:
Version can be null for now

haacked committed:
Fix a bunch of tests and snapshots

haacked committed:
Include `last_modified_by` in query.

haacked committed:
Remove unnecessary and incorrect values

haacked committed:
Clean up code

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

haacked committed:
Reload flags when switching to edit mode

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

haacked committed:
Fix up more tests

haacked committed:
Fix up more tests

haacked committed:
Remove blank=True

It's unnecessary.

github-actions[bot] committed:
Update query snapshots

havenbarnes commented on posthog/api/feature_flag.py:
```diff
@@ -68,6 +70,9 @@
 from ee.models.rbac.organization_resource_access import OrganizationResourceAccess
+import structlog
+
+logger = structlog.get_logger(__name__)
```
nit: I don't think this is used

haacked replied:
Oh yeah, I can remove it. I think I planned to log when we had a conflict so we can get a sense of how often this happens. But I'm wondering if this should be a captured event or something else? Or not worth logging at all? Thoughts?

havenbarnes reviewed (approved):
Very cool! Big improvement

haacked committed:
Remove unused logger

haacked committed:
Don't automatically create index

Creating the index automatically triggers this error in CI:
Found a CREATE INDEX command that isn't run CONCURRENTLY. This locks tables which causes downtime.

haacked committed:
Update migrations

haacked committed:
Fix model to match migration

haacked committed:
We don't filter by these properties.

haacked committed:
Remove migration.

haacked committed:
Merge branch 'master' into haacked/17995-concurrency

haacked committed:
Rebuild Migration

haacked committed:
Format with ruff .
